### PR TITLE
wrote new controller

### PR DESCRIPTION
**Pull request by Zingerd**

### CODE REVIEW

**Review by Natalia1910 — APPROVED** [review]

**Review by dessdee — APPROVED** [review]

**Review by Max877866 — APPROVED** [review]

File: core/src/main/java/greencity/controller/OrderController.java [hunk 1]
```diff
@@ -490,4 +490,27 @@ public OrderController(UBSClientService ubsClientService) {
...
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = HttpStatuses.OK),
+        @ApiResponse(code = 400, message = HttpStatuses.BAD_REQUEST)
+    })
+    @PostMapping("/receivePaymentClient")
+    public ResponseEntity<HttpStatus> receivePaymentClient(
+        PaymentResponseDto dto, HttpServletResponse response) throws IOException {
+        ubsClientService.validatePaymentClient(dto);
+        if (HttpStatus.OK.is2xxSuccessful()) {
+            response.sendRedirect("https://ita-social-projects.github.io/GreenCityClient/#/ubs/confirm");
```
Comment (thesvider) [inline core/src/main/java/greencity/controller/OrderController.java:512]: https://ita-social-projects.github.io/GreenCityClient/#/ubs/confirm - should be in properties.
https://ita-social-projects.github.io/GreenCityClient - should be replaced with some ${base_url} parameter.

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 2]
```diff
@@ -1344,14 +1326,16 @@ private Order incrementCounter(Order order) {
     private PaymentRequestDto formPayment(Long orderId, int sumToPay) {
         Order order = orderRepository.findById(orderId)
             .orElseThrow(() -> new OrderNotFoundException(ORDER_WITH_CURRENT_ID_DOES_NOT_EXIST));
+        int lastNumber = order.getPayment().size() - 1;
         PaymentRequestDto paymentRequestDto = PaymentRequestDto.builder()
             .merchantId(Integer.parseInt(merchantId))
             .orderId(
-                orderId + order.getCounterOrderPaymentId().toString() + order.getPayment().get(0).getId())
+                orderId + "_" + order.getCounterOrderPaymentId().toString() + "_"
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1333]: String.format("%s_%s_%s", orderId, order.getCounterOrderPaymentId(), order.getPayment().get(lastNumber).getId())

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 3]
```diff
@@ -1344,14 +1326,16 @@ private Order incrementCounter(Order order) {
...
+                    + order.getPayment().get(lastNumber).getId())
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1334]: make sure order.getPayment().get(lastNumber).getId() is NPE safe and no out of bound possible

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 4]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+        Order order = orderRepository.findById(Long.valueOf(id[0]))
+            .orElseThrow(() -> new PaymentValidationException(PAYMENT_VALIDATION_ERROR));
+
+        checkResponseStatusFailure(dto, orderPayment, order);
+        checkResponseValidationSignature(dto);
+        checkOrderStatusApproved(dto, orderPayment, order);
+    }
+
+    private Payment mapPaymentClient(PaymentResponseDto dto) {
+        String[] idClient = orderIdInfo(dto);
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1518]: OrderUtils:

// match with order id from payment response using regexp
 

public static String generateOrderIdForPayment(Order order);
public static Long extractOrderIdFromPaymentResponse(String);

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 5]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+            .responseStatus(dto.getResponse_status())
+            .senderCellPhone(dto.getSender_cell_phone())
+            .senderAccount(dto.getSender_account())
+            .maskedCard(dto.getMasked_card())
+            .cardType(dto.getCard_type())
+            .responseCode(dto.getResponse_code())
+            .responseDescription(dto.getResponse_description())
+            .orderTime(dto.getOrder_time())
+            .settlementDate(dto.getSettlement_date())
+            .fee(Long.valueOf(dto.getFee()))
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1539]: Optional.ofNullable(dto.getFee())map(Long::valueOf).orElse(0L)

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 6]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+            .paymentSystem(dto.getPayment_system())
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1540]: @JsonProperty("payment_system")
String paymentSystem;

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 7]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+            .paymentStatus(PaymentStatus.UNPAID)
+            .build();
+    }
+
+    private String[] orderIdInfo(PaymentResponseDto dto) {
+        return dto.getOrder_id().split("_");
+    }
+
+    private void checkResponseStatusFailure(PaymentResponseDto dto, Payment orderPayment, Order order) {
+        if (dto.getResponse_status().equals("failure")) {
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1552]: PaymentResponseDto
private static final String FAILED_STATUS = "failure";

boolean isFailed() {
returnFAILED_STATUS .equals(reponseStatus)
}

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 8]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+    }
+
+    private void checkResponseValidationSignature(PaymentResponseDto dto) {
+        if (!encryptionUtil.checkIfResponseSignatureIsValid(dto, fondyPaymentKey)) {
+            throw new PaymentValidationException(PAYMENT_VALIDATION_ERROR);
+        }
+    }
+
+    private void checkOrderStatusApproved(PaymentResponseDto dto, Payment orderPayment, Order order) {
+        if (dto.getOrder_status().equals("approved")) {
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1567]: PaymentResponseDto
private static final String APPROVED_STATUS = "approved";

boolean isApproved() {
return APPROVED_STATUS .equals(reponseStatus)
}

**Review by thesvider — COMMENTED** [review]
please resolve comments in next PR

File: service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java [hunk 9]
```diff
@@ -1516,4 +1500,86 @@ private PaymentRequestDtoLiqPay formLiqPayPaymentForIF(Long orderId, int sumToPa
...
+    }
+
+    private Payment mapPaymentClient(PaymentResponseDto dto) {
+        String[] idClient = orderIdInfo(dto);
+        Order order = orderRepository.findById(Long.valueOf(idClient[0]))
+            .orElseThrow(() -> new PaymentValidationException(PAYMENT_VALIDATION_ERROR));
+        int lastNumber = order.getPayment().size() - 1;
+        checkDtoFee(dto);
+
+        return Payment.builder()
```
Comment (thesvider) [inline service/src/main/java/greencity/service/ubs/UBSClientServiceImpl.java:1524]: please check if this block can be extracted into new mapper class

**Pull request merged**